dags/operators/emr_spark_operator.py

```diff
@@ -0,0 +1,138 @@
+import logging
+import boto3
+import requests
+import time
+
+from airflow.models import BaseOperator
+from airflow.utils import apply_defaults, AirflowException
+from os import environ
+
+class EMRSparkOperator(BaseOperator):
+    """
+    Execute a Spark job on EMR.
+
+    :param job_name: The name of the job.
+    :type job_name: string
+
+    :param owner: The e-mail address of the user owning the job.
+    :type owner: string
+
+    :param uri: The URI of the job to run, which can be either a Jupyter notebook
+                or a JAR file.
+    :type uri: string
+
+    :param instance_count: The number of workers the cluster should have.
+    :type instance_count: int
+
+    :param env: If env is not None, it must be a mapping that defines the environment
+                variables for the new process (templated).
+    :type env: string
+    """
+    template_fields = ('environment', )
+
+    region = environ["REGION"]
+    key_name = environ["EMR_KEY_NAME"]
+    release_label = environ["EMR_RELEASE_LABEL"]
+    flow_role = environ["EMR_FLOW_ROLE"]
+    service_role = environ["EMR_SERVICE_ROLE"]
+    instance_type = environ["EMR_INSTANCE_TYPE"]
+    spark_bucket = environ["SPARK_BUCKET"]
+    airflow_bucket = environ["AIRFLOW_BUCKET"]
+
+    def __del__(self):
+        self.on_kill()
+
+
+    def post_execute(self, context):
+        self.on_kill()
+
+
+    def on_kill(self):
+        if self.job_flow_id is None:
+            return
+
+        client = boto3.client('emr', region_name=EMRSparkOperator.region)
+        result = client.describe_cluster(ClusterId = self.job_flow_id)
+        status = result["Cluster"]["Status"]["State"]
+
+        if status != "TERMINATED_WITH_ERRORS" and status != "TERMINATED":
+            logging.warn("Terminating Spark job {}".format(self.job_name))
+            client.terminate_job_flows(JobFlowIds = [self.job_flow_id])
+
+
+    @apply_defaults
+    def __init__(self, job_name, owner, uri, instance_count, env={}, arguments="", *args, **kwargs):
+        super(EMRSparkOperator, self).__init__(*args, **kwargs)
+        self.job_name = job_name
+        self.owner = owner
+        self.uri = uri
+        self.arguments = arguments
+        self.environment = " ".join(["{}={}".format(k, v) for k, v in env.iteritems()])
```

Comment:
We should probably validate that the items in `env` are not going to cause problems (by containing spaces or quotes etc) and throw an error